File: crates/breez-sdk/bindings/examples/cli/langs/go/issuer.go
package main

import (
	"flag"
	"fmt"
	"math/big"
	"sort"

	breez_sdk_spark "github.com/breez/breez-sdk-spark-go/breez_sdk_spark"
	"github.com/chzyer/readline"
)

// IssuerCommand represents a single issuer subcommand.
type IssuerCommand struct {
	Name        string
	Description string
	Run         func(issuer *breez_sdk_spark.TokenIssuer, rl *readline.Instance, args []string) error
}

// IssuerCommandNames lists all issuer subcommand names (used for REPL completion).
var IssuerCommandNames = []string{
	"issuer token-balance",
	"issuer token-metadata",
	"issuer create-token",
	"issuer mint-token",
	"issuer burn-token",
	"issuer freeze-token",
	"issuer unfreeze-token",
}

// BuildIssuerRegistry returns a map of issuer subcommand name → IssuerCommand.
func BuildIssuerRegistry() map[string]IssuerCommand {
	return map[string]IssuerCommand{
		"token-balance": {
			Name:        "token-balance",
			Description: "Get issuer token balance",
			Run:         handleTokenBalance,
		},
		"token-metadata": {
			Name:        "token-metadata",
			Description: "Get issuer token metadata",
			Run:         handleTokenMetadata,
		},
		"create-token": {
			Name:        "create-token",
			Description: "Create a new issuer token",
			Run:         handleCreateToken,
		},
		"mint-token": {
			Name:        "mint-token",
			Description: "Mint supply of the issuer token",
			Run:         handleMintToken,
		},
		"burn-token": {
			Name:        "burn-token",
			Description: "Burn supply of the issuer token",
			Run:         handleBurnToken,
		},
		"freeze-token": {
			Name:        "freeze-token",
			Description: "Freeze tokens at an address",
			Run:         handleFreezeToken,
		},
		"unfreeze-token": {
			Name:        "unfreeze-token",
			Description: "Unfreeze tokens at an address",
			Run:         handleUnfreezeToken,
		},
	}
}

// DispatchIssuerCommand dispatches an issuer subcommand.
func DispatchIssuerCommand(args []string, issuer *breez_sdk_spark.TokenIssuer, rl *readline.Instance) {
	registry := BuildIssuerRegistry()

	if len(args) == 0 || args[0] == "help" {
		fmt.Println("\nIssuer subcommands:\n")
		names := make([]string, 0, len(registry))
		for name := range registry {
			names = append(names, name)
		}
		sort.Strings(names)
		for _, name := range names {
			cmd := registry[name]
			fmt.Printf("  issuer %-30s %s\n", name, cmd.Description)
		}
		fmt.Println()
		return
	}

	subName := args[0]
	subArgs := args[1:]

	cmd, ok := registry[subName]
	if !ok {
		fmt.Printf("Unknown issuer subcommand: %s. Use 'issuer help' for available commands.\n", subName)
		return
	}

	if err := cmd.Run(issuer, rl, subArgs); err != nil {
		fmt.Printf("Error: %v\n", err)
	}
}

// ---------------------------------------------------------------------------
// Issuer command handlers
// ---------------------------------------------------------------------------

// --- token-balance ---

func handleTokenBalance(issuer *breez_sdk_spark.TokenIssuer, _ *readline.Instance, _ []string) error {
	result, err := issuer.GetIssuerTokenBalance()
	if err = liftError(err); err != nil {
		return err
	}
	printValue(result)
	return nil
}

// --- token-metadata ---

func handleTokenMetadata(issuer *breez_sdk_spark.TokenIssuer, _ *readline.Instance, _ []string) error {
	result, err := issuer.GetIssuerTokenMetadata()
	if err = liftError(err); err != nil {
		return err
	}
	printValue(result)
	return nil
}

// --- create-token ---

func handleCreateToken(issuer *breez_sdk_spark.TokenIssuer, _ *readline.Instance, args []string) error {
	fs := flag.NewFlagSet("create-token", flag.ContinueOnError)
	name := fs.String("name", "", "Token name (required)")
	ticker := fs.String("ticker", "", "Token ticker (required)")
	decimals := fs.Uint("decimals", 6, "Token decimals")
	freezable := fs.Bool("freezable", false, "Whether the token can be frozen")
	maxSupplyStr := fs.String("max-supply", "", "Maximum supply (optional)")
	if err := fs.Parse(args); err != nil {
		return err
	}

	if *name == "" || *ticker == "" {
		fmt.Println("Usage: issuer create-token --name <name> --ticker <ticker> [--decimals N] [--freezable] [--max-supply N]")
		return nil
	}

	decimalsU32 := uint32(*decimals)
	req := breez_sdk_spark.CreateIssuerTokenRequest{
		Name:        *name,
		Ticker:      *ticker,
		Decimals:    decimalsU32,
		IsFreezable: *freezable,
	}

	if *maxSupplyStr != "" {
		maxSupply, ok := new(big.Int).SetString(*maxSupplyStr, 10)
		if !ok {
			return fmt.Errorf("invalid max-supply: %s", *maxSupplyStr)
		}
		req.MaxSupply = maxSupply
	}

	result, err := issuer.CreateIssuerToken(req)
	if err = liftError(err); err != nil {
		return err
	}
	printValue(result)
	return nil
}

// --- mint-token ---

func handleMintToken(issuer *breez_sdk_spark.TokenIssuer, _ *readline.Instance, args []string) error {
	if len(args) < 1 {
		fmt.Println("Usage: issuer mint-token <amount>")
		return nil
	}

	amount, ok := new(big.Int).SetString(args[0], 10)
	if !ok {
		return fmt.Errorf("invalid amount: %s", args[0])
	}

	result, err := issuer.MintIssuerToken(breez_sdk_spark.MintIssuerTokenRequest{
		Amount: amount,
	})
	if err = liftError(err); err != nil {
		return err
	}
	printValue(result)
	return nil
}

// --- burn-token ---

func handleBurnToken(issuer *breez_sdk_spark.TokenIssuer, _ *readline.Instance, args []string) error {
	if len(args) < 1 {
		fmt.Println("Usage: issuer burn-token <amount>")
		return nil
	}

	amount, ok := new(big.Int).SetString(args[0], 10)
	if !ok {
		return fmt.Errorf("invalid amount: %s", args[0])
	}

	result, err := issuer.BurnIssuerToken(breez_sdk_spark.BurnIssuerTokenRequest{
		Amount: amount,
	})
	if err = liftError(err); err != nil {
		return err
	}
	printValue(result)
	return nil
}

// --- freeze-token ---

func handleFreezeToken(issuer *breez_sdk_spark.TokenIssuer, _ *readline.Instance, args []string) error {
	if len(args) < 1 {
		fmt.Println("Usage: issuer freeze-token <address>")
		return nil
	}

	result, err := issuer.FreezeIssuerToken(breez_sdk_spark.FreezeIssuerTokenRequest{
		Address: args[0],
	})
	if err = liftError(err); err != nil {
		return err
	}
	printValue(result)
	return nil
}

// --- unfreeze-token ---

func handleUnfreezeToken(issuer *breez_sdk_spark.TokenIssuer, _ *readline.Instance, args []string) error {
	if len(args) < 1 {
		fmt.Println("Usage: issuer unfreeze-token <address>")
		return nil
	}

	result, err := issuer.UnfreezeIssuerToken(breez_sdk_spark.UnfreezeIssuerTokenRequest{
		Address: args[0],
	})
	if err = liftError(err); err != nil {
		return err
	}
	printValue(result)
	return nil
}

File: crates/breez-sdk/bindings/examples/cli/langs/csharp/Issuer.cs
using System.Numerics;
using Breez.Sdk.Spark;

namespace BreezCli;

/// <summary>
/// Represents a single issuer subcommand.
/// </summary>
public class IssuerCliCommand
{
    public required string Name { get; init; }
    public required string Description { get; init; }
    public required Func<TokenIssuer, Func<string, string?>, string[], Task> Run { get; init; }
}

/// <summary>
/// Issuer subcommand names (used for REPL completion).
/// </summary>
public static class IssuerCommandNames
{
    public static readonly string[] All =
    {
        "issuer token-balance",
        "issuer token-metadata",
        "issuer create-token",
        "issuer mint-token",
        "issuer burn-token",
        "issuer freeze-token",
        "issuer unfreeze-token",
    };
}

/// <summary>
/// Token issuer subcommand handlers.
/// </summary>
public static class IssuerCommands
{
    /// <summary>
    /// Builds the issuer subcommand registry.
    /// </summary>
    public static Dictionary<string, IssuerCliCommand> BuildRegistry()
    {
        return new Dictionary<string, IssuerCliCommand>
        {
            ["token-balance"] = new()
            {
                Name = "token-balance",
                Description = "Get issuer token balance",
                Run = HandleTokenBalance
            },
            ["token-metadata"] = new()
            {
                Name = "token-metadata",
                Description = "Get issuer token metadata",
                Run = HandleTokenMetadata
            },
            ["create-token"] = new()
            {
                Name = "create-token",
                Description = "Create a new issuer token",
                Run = HandleCreateToken
            },
            ["mint-token"] = new()
            {
                Name = "mint-token",
                Description = "Mint supply of the issuer token",
                Run = HandleMintToken
            },
            ["burn-token"] = new()
            {
                Name = "burn-token",
                Description = "Burn supply of the issuer token",
                Run = HandleBurnToken
            },
            ["freeze-token"] = new()
            {
                Name = "freeze-token",
                Description = "Freeze tokens at an address",
                Run = HandleFreezeToken
            },
            ["unfreeze-token"] = new()
            {
                Name = "unfreeze-token",
                Description = "Unfreeze tokens at an address",
                Run = HandleUnfreezeToken
            },
        };
    }

    /// <summary>
    /// Dispatches an issuer subcommand.
    /// </summary>
    public static async Task DispatchCommand(
        string[] args,
        TokenIssuer tokenIssuer,
        Func<string, string?> readline)
    {
        var registry = BuildRegistry();

        if (args.Length == 0 || args[0] == "help")
        {
            Console.WriteLine();
            Console.WriteLine("Issuer subcommands:");
            var names = registry.Keys.OrderBy(k => k).ToList();
            foreach (var name in names)
            {
                Console.WriteLine($"  issuer {name,-30} {registry[name].Description}");
            }
            Console.WriteLine();
            return;
        }

        var subName = args[0];
        var subArgs = args.Skip(1).ToArray();

        if (!registry.TryGetValue(subName, out var cmd))
        {
            Console.WriteLine($"Unknown issuer subcommand: {subName}. Use 'issuer help' for available commands.");
            return;
        }

        await cmd.Run(tokenIssuer, readline, subArgs);
    }

    // -----------------------------------------------------------------------
    // Argument parsing helpers
    // -----------------------------------------------------------------------

    private static string? GetFlag(string[] args, params string[] names)
    {
        for (int i = 0; i < args.Length - 1; i++)
        {
            if (names.Contains(args[i]))
            {
                return args[i + 1];
            }
        }
        return null;
    }

    private static bool HasFlag(string[] args, params string[] names)
    {
        return args.Any(a => names.Contains(a));
    }

    // -----------------------------------------------------------------------
    // Issuer command handlers
    // -----------------------------------------------------------------------

    // --- token-balance ---

    private static async Task HandleTokenBalance(TokenIssuer issuer, Func<string, string?> readline, string[] args)
    {
        var result = await issuer.GetIssuerTokenBalance();
        Serialization.PrintValue(result);
    }

    // --- token-metadata ---

    private static async Task HandleTokenMetadata(TokenIssuer issuer, Func<string, string?> readline, string[] args)
    {
        var result = await issuer.GetIssuerTokenMetadata();
        Serialization.PrintValue(result);
    }

    // --- create-token ---

    private static async Task HandleCreateToken(TokenIssuer issuer, Func<string, string?> readline, string[] args)
    {
        var name = GetFlag(args, "--name");
        var ticker = GetFlag(args, "--ticker");
        var decimalsStr = GetFlag(args, "--decimals") ?? "6";
        var freezable = HasFlag(args, "-f", "--freezable");
        var maxSupplyStr = GetFlag(args, "--max-supply");

        if (name == null || ticker == null)
        {
            Console.WriteLine("Usage: issuer create-token --name <name> --ticker <ticker> [--decimals N] [--freezable] [--max-supply N]");
            return;
        }

        var decimals = uint.Parse(decimalsStr);
        BigInteger maxSupply = maxSupplyStr != null ? BigInteger.Parse(maxSupplyStr) : BigInteger.Zero;

        var result = await issuer.CreateIssuerToken(new CreateIssuerTokenRequest(
            name: name,
            ticker: ticker,
            decimals: decimals,
            isFreezable: freezable,
            maxSupply: maxSupply
        ));
        Serialization.PrintValue(result);
    }

    // --- mint-token ---

    private static async Task HandleMintToken(TokenIssuer issuer, Func<string, string?> readline, string[] args)
    {
        if (args.Length < 1)
        {
            Console.WriteLine("Usage: issuer mint-token <amount>");
            return;
        }

        var amount = BigInteger.Parse(args[0]);
        var result = await issuer.MintIssuerToken(new MintIssuerTokenRequest(amount: amount));
        Serialization.PrintValue(result);
    }

    // --- burn-token ---

    private static async Task HandleBurnToken(TokenIssuer issuer, Func<string, string?> readline, string[] args)
    {
        if (args.Length < 1)
        {
            Console.WriteLine("Usage: issuer burn-token <amount>");
            return;
        }

        var amount = BigInteger.Parse(args[0]);
        var result = await issuer.BurnIssuerToken(new BurnIssuerTokenRequest(amount: amount));
        Serialization.PrintValue(result);
    }

    // --- freeze-token ---

    private static async Task HandleFreezeToken(TokenIssuer issuer, Func<string, string?> readline, string[] args)
    {
        if (args.Length < 1)
        {
            Console.WriteLine("Usage: issuer freeze-token <address>");
            return;
        }

        var result = await issuer.FreezeIssuerToken(new FreezeIssuerTokenRequest(address: args[0]));
        Serialization.PrintValue(result);
    }

    // --- unfreeze-token ---

    private static async Task HandleUnfreezeToken(TokenIssuer issuer, Func<string, string?> readline, string[] args)
    {
        if (args.Length < 1)
        {
            Console.WriteLine("Usage: issuer unfreeze-token <address>");
            return;
        }

        var result = await issuer.UnfreezeIssuerToken(new UnfreezeIssuerTokenRequest(address: args[0]));
        Serialization.PrintValue(result);
    }
}
